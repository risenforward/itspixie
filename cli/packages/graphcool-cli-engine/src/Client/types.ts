export interface User {
  id: string
  name: string
  login: Login[]
}

export interface Login {
  email: string
}
