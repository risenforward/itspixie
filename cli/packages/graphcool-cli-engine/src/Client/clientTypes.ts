export interface DeployPayload {
  project: SimpleProject
  errors: SchemaError[]
  migration: Migration[]
}

export interface SchemaError {
  type: string
  field?: string
  description: string
}

export interface SimpleProject {
  id: string
}

export interface Migration {
  revision: number
  hasBeenApplied: boolean
  steps: MigrationStep[]
}

export interface MigrationStep {
  type: string
  name: string
  // createEnum
  ce_values?: string
  // createField
  cf_typeName?: string
  cf_isRequired?: boolean
  cf_isList?: boolean
  cf_isUnique?: boolean
  cf_relation?: string
  cf_defaultValue?: string
  cf_enum?: string
  // createRelation
  leftModel?: string
  rightModel?: string
  // deleteField
  model?: string
  // updateEnum
  newName?: string
  values?: string[]
  // updateField
  typeName?: string
  isRequried?: boolean
  isList?: boolean
  isRequired?: boolean
  relation?: string
  defaultValue?: string
  enum?: string
  // updateModel
  um_newName: string
}
