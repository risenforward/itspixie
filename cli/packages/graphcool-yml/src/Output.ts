export class Output {
  warn(...args) {
    console.warn(args)
  }
  getErrorPrefix(fileName: string, type: 'error' | 'warning' = 'error') {
    return `[${type.toUpperCase()}] in ${fileName}: `
  }
}

export interface IOutput {
  warn: (...args) => void
  getErrorPrefix: (fileName: string, type?: 'error' | 'warning') => string
}
