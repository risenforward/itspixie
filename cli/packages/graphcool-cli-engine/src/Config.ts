import { RunOptions } from './types'
import * as path from 'path'
import * as os from 'os'
import * as fs from 'fs-extra'
import * as cuid from 'cuid'
const debug = require('debug')('config')
import * as findUp from 'find-up'
import { Output } from './Output/index'

export class Config {
  /**
   * Local settings
   */
  out: Output
  debug: boolean = Boolean(
    process.env.DEBUG && process.env.DEBUG!.includes('*'),
  )
  windows: boolean = false
  bin: string = 'graphcool'
  mock: boolean = true
  argv: string[] = process.argv.slice(1)
  commandsDir: string = path.join(__dirname, '../dist/commands')
  defaultCommand: string = 'help'
  userPlugins: boolean = false
  version: string = '1.3.11'
  name: string = 'graphcool'
  pjson: any = {
    name: 'cli-engine',
    version: '0.0.0',
    dependencies: {},
    'cli-engine': {
      defaultCommand: 'help',
    },
  }

  /**
   * Paths
   */
  cwd: string
  home: string
  root = path.join(__dirname, '..')

  envPath: string | null
  definitionDir: string
  definitionPath: string | null
  dotGraphcoolFilePath: string | null
  warnings: string[] = []

  /**
   * Urls
   */
  token: string | null
  authUIEndpoint = process.env.ENV === 'DEV'
    ? 'https://dev.console.graph.cool/cli/auth'
    : 'https://console.graph.cool/cli/auth'
  backendAddr = process.env.ENV === 'DEV'
    ? 'https://dev.api.graph.cool'
    : 'https://api.graph.cool'
  systemAPIEndpoint = process.env.ENV === 'DEV'
    ? 'https://dev.api.graph.cool/system'
    : 'https://api.graph.cool/system'
  authEndpoint = process.env.ENV === 'DEV'
    ? 'https://cli-auth-api.graph.cool/dev'
    : 'https://cli-auth-api.graph.cool/prod'
  docsEndpoint = process.env.ENV === 'DEV'
    ? 'https://dev.graph.cool/docs'
    : 'https://www.graph.cool/docs'
  statusEndpoint = 'https://crm.graph.cool/prod/status'

  /* tslint:disable-next-line */
  __cache = {}

  constructor(options?: RunOptions) {
    this.cwd = (options && options.cwd) || this.getCwd()
    this.home = (options && options.home) || this.getHome()
    debug(`CWD`, this.cwd)
    debug(`HOME`, this.home)
    this.setDotGraphcoolPath()
    this.setEnvPath()
    this.setDefinitionPaths()
    debug(`dotGraphcoolPath after setting it`, this.dotGraphcoolFilePath)
    this.setTokenIfExists()
    if (options) {
      this.readPackageJson(options)
    }
  }
  setLocal(host: string = 'http://localhost:60000') {
    this.backendAddr = host
    this.systemAPIEndpoint = host + '/system'
  }
  setOutput(out: Output) {
    this.out = out
    this.warnings.forEach(warning  => out.warn(warning))
    this.warnings = []
  }
  setToken(token: string | null) {
    this.token = token
  }
  saveToken() {
    if (this.dotGraphcoolFilePath) {
      const json = JSON.stringify({ token: this.token }, null, 2)
      fs.writeFileSync(this.dotGraphcoolFilePath, json)
    }
  }
  get arch(): string {
    return os.arch() === 'ia32' ? 'x86' : os.arch()
  }
  get platform(): string {
    return os.platform() === 'win32' ? 'windows' : os.platform()
  }
  get userAgent(): string {
    return `${this.name}/${this.version} (${this.platform}-${this
      .arch}) node-${process.version}`
  }
  get dirname() {
    return this.pjson['cli-engine'].dirname || this.bin
  }
  get cacheDir() {
    return dir(
      this,
      'cache',
      this.platform === 'darwin'
        ? path.join(this.home, 'Library', 'Caches')
        : null,
    )
  }
  private readPackageJson(options: RunOptions) {
    this.mock = options.mock
    this.argv = options.argv || this.argv
    if (options.root) {
      this.root = options.root
      const pjsonPath = path.join(options.root, 'package.json')
      const pjson = fs.readJSONSync(pjsonPath)
      if (pjson && pjson['cli-engine']) {
        this.pjson = pjson
        this.version = pjson.version
      }
    }
  }
  private setEnvPath() {
    this.envPath = path.join(this.cwd, '.graphcoolrc')
    if (!fs.pathExistsSync(this.envPath)) {
      const found = findUp.sync('.graphcoolrc', {cwd: this.cwd})
      if (found) {

        // only use this if 1. it's not in the home dir (people of old cli versions may have this)
        // and 2. there MUST be a graphcool.yml file in the same folder, otherwise this makes no sense to use!
        const foundDir = path.dirname(found)

        if (foundDir === this.home) {
          const file = fs.readFileSync(found, 'utf-8')
          let json: any = null
          try {
            json = JSON.parse(file)
          } catch (e) {
            //
          }

          if (json && json.token && (!this.dotGraphcoolFilePath || !fs.pathExistsSync(this.dotGraphcoolFilePath))) {
            const newDotPath = path.join(this.home, '.graphcool')
            fs.moveSync(found, newDotPath)
            this.dotGraphcoolFilePath = newDotPath
          } else {
            this.warn(`There is a .graphcoolrc file in your home directory (${this.envPath}).
This can still be an artifact of the old CLI version.
To prevent unwanted side effects, please remove it.`)
          }

          return
        }

        const ymlPath = path.join(foundDir, 'graphcool.yml')
        if (fs.pathExistsSync(ymlPath)) {
          this.envPath = (found && path.dirname(found) !== this.home) ? found : this.envPath
        }
      }
    }
  }
  private warn(msg: string) {
    this.warnings.push(msg)
  }
  private setDefinitionPaths() {
    const definitionPath = path.join(this.cwd, 'graphcool.yml')
    if (fs.pathExistsSync(definitionPath)) {
      this.definitionDir = this.cwd
      this.definitionPath = definitionPath
    } else {
      const found = findUp.sync('graphcool.yml', {cwd: this.cwd})
      this.definitionDir = found ? path.dirname(found) : this.cwd
      this.definitionPath = found || null
    }
    debug(`definitionDir`, this.definitionDir)
    debug(`definitionPath`, this.definitionPath)
  }
  private setDotGraphcoolPath() {
    const dotGraphcoolCwd = path.join(this.cwd, '.graphcool')
    if (fs.pathExistsSync(dotGraphcoolCwd)) {
      this.dotGraphcoolFilePath = dotGraphcoolCwd
    } else {
      const found = findUp.sync('.graphcool', {cwd: this.cwd})
      const dotGraphcoolHome = path.join(this.home, '.graphcool')

      // only take the find-up file, if it's "deeper" than the home dir
      this.dotGraphcoolFilePath = (found && (found.split('/').length > dotGraphcoolHome.split('/'))) ? found : dotGraphcoolHome
    }
  }
  private setTokenIfExists() {
    if (process.env.NODE_ENV === 'test') {
      debug('taking graphcool test token')
      this.token = process.env.GRAPHCOOL_TEST_TOKEN!
    } else {
      if (this.dotGraphcoolFilePath && fs.existsSync(this.dotGraphcoolFilePath)) {
        const configContent = fs.readFileSync(
          this.dotGraphcoolFilePath,
          'utf-8',
        )
        this.token = JSON.parse(configContent).token
      }
    }
  }
  private getCwd() {
    // get cwd
    let cwd = process.cwd()
    if (process.env.NODE_ENV === 'test') {
      cwd = path.join(os.tmpdir(), `${cuid()}/`)
      fs.mkdirpSync(cwd)
      debug('cwd', cwd)
    }
    return cwd
  }
  private getHome() {
    // get home
    let home = os.homedir() || os.tmpdir()
    if (process.env.NODE_ENV === 'test') {
      home = path.join(os.tmpdir(), `${cuid()}/`)
      fs.mkdirpSync(home)
      debug('home', home)
    }
    return home
  }
}

function dir(config: Config, category: string, d: string | null): string {
  const cacheKey = `dir:${category}`
  const cache = config.__cache[cacheKey]
  if (cache) {
    return cache
  }
  d =
    d ||
    path.join(
      config.home,
      category === 'data' ? '.local/share' : '.' + category,
    )
  if (config.windows) {
    d = process.env.LOCALAPPDATA || d
  }
  d = process.env.XDG_DATA_HOME || d
  d = path.join(d, config.dirname)
  fs.mkdirpSync(d)
  config.__cache[cacheKey] = d
  return d
}
