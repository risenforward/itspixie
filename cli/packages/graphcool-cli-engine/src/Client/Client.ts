import {
  AccountInfo,
  AuthenticateCustomerPayload,
  FunctionInfo,
  FunctionLog,
  PAT,
  Project,
  ProjectDefinition,
  ProjectInfo,
  RemoteProject,
  SimpleProjectInfo,
} from '../types/common'

import { GraphQLClient, request } from 'graphql-request'
import { omit, flatMap } from 'lodash'
import { Config } from '../Config'
import { getFastestRegion } from './ping'
import { Environment, Cluster } from 'graphcool-yml'
import { Output } from '../index'
import chalk from 'chalk'
import { DeployPayload } from './clientTypes'
import { introspectionQuery } from './introspectionQuery'

const debug = require('debug')('client')

const REMOTE_PROJECT_FRAGMENT = `
  fragment RemoteProject on Project {
    id
    name
    schema
    alias
    region
    isEjected
    projectDefinitionWithFileContent
  }
`

const MIGRATION_FRAGMENT = `
fragment MigrationFragment on Migration {
  revision
  steps {
    type
    __typename
    ... on CreateEnum {
      name
      ce_values: values
    }
    ... on CreateField {
      model
      name
      cf_typeName: typeName
      cf_isRequired: isRequired
      cf_isList: isList
      cf_isUnique: unique
      cf_relation: relation
      cf_defaultValue: default
      cf_enum: enum
    }
    ... on CreateModel {
      name
    }
    ... on CreateRelation {
      name
      leftModel
      rightModel
    }
    ... on DeleteEnum {
      name
    }
    ... on DeleteField {
      model
      name
    }
    ... on DeleteModel {
      name
    }
    ... on DeleteRelation {
      name
    }
    ... on UpdateEnum {
      name
      newName
      values
    }
    ... on UpdateField {
      model
      name
      newName
      typeName
      isRequired
      isList
      isUnique: unique
      relation
      default
      enum
    }
    ... on UpdateModel {
      name
      um_newName: newName
    }
  }
  hasBeenApplied
}
`

export class Client {
  config: Config
  env: Environment
  out: Output
  public mock: (input: { request: any; response: any }) => void

  private mocks: { [request: string]: string } = {}
  private tokenCache: string

  constructor(config: Config, environment: Environment, out: Output) {
    this.config = config
    this.env = environment
    this.out = out
  }

  // always create a new client which points to the latest config for each request
  getClient(cluster: Cluster) {
    return new GraphQLClient(cluster.getDeployEndpoint(), {
      headers: {
        Authorization: `Bearer ${cluster.token}`,
      },
    })
  }
  get client(): GraphQLClient {
    if (!this.env.activeCluster) {
      throw new Error(
        `No cluster set. Please set the "cluster" property in your graphcool.yml`,
      )
    }
    debug(
      'choosing clusterEndpoint',
      this.env.activeCluster.getDeployEndpoint(),
    )
    const localClient = this.getClient(this.env.activeCluster!)
    return {
      request: async (query, variables) => {
        debug('Sending query')
        debug(query)
        debug(variables)
        try {
          return await localClient.request(query, variables)
        } catch (e) {
          if (e.message.startsWith('No valid session')) {
            // await this.auth.ensureAuth(true)
            // try again with new token
            return await this.client.request(query, variables)
          } else if (
            e.message.includes('ECONNREFUSED') &&
            (e.message.includes('localhost') || e.message.includes('127.0.0.1'))
          ) {
            this.out.error(
              `Could not connect to local cluster. Please use ${chalk.bold.green(
                'graphcool local start',
              )} to start your local Graphcool cluster.`,
            )
          } else {
            throw e
          }
        }
      },
    } as any
  }

  // async getAccount(): Promise<AccountInfo> {
  //   const { viewer: { user } } = await this.client.request<{
  //     viewer: { user: AccountInfo }
  //   }>(`{
  //     viewer {
  //       user {
  //         email
  //         name
  //       }
  //     }
  //   }`)

  //   return user
  // }

  async introspect(
    serviceName: string,
    stageName: string,
    token?: string,
  ): Promise<any> {
    debug('introspecting', serviceName, stageName, token)
    const headers: any = {}
    if (token) {
      headers.Authorization = `Bearer ${token}`
    }
    const client = new GraphQLClient(
      this.env.activeCluster.getApiEndpoint(serviceName, stageName),
      {
        headers,
      },
    )
    return client.request(introspectionQuery)
  }

  async exec(
    serviceName: string,
    stageName: string,
    query: string,
    token?: string,
  ): Promise<any> {
    debug('executing query', serviceName, stageName, query, token)
    const headers: any = {}
    if (token) {
      headers.Authorization = `Bearer ${token}`
    }
    const client = new GraphQLClient(
      this.env.activeCluster.getApiEndpoint(serviceName, stageName),
      {
        headers,
      },
    )
    return client.request(query)
  }

  async download(
    serviceName: string,
    stage: string,
    exportData: any,
    token?: string,
  ): Promise<any> {
    const endpoint = this.env.activeCluster.getExportEndpoint(
      serviceName,
      stage,
    )
    debug(`Downloading from ${endpoint}`)
    debug(exportData)
    const result = await fetch(endpoint, {
      method: 'post',
      headers: {
        Authorization: `Bearer ${token}`,
        'Content-Type': 'application/json',
      },
      body: exportData,
    })

    const text = await result.text()
    try {
      return JSON.parse(text)
    } catch (e) {
      throw new Error(text)
    }
  }

  async upload(
    serviceName: string,
    stage: string,
    exportData: any,
    token?: string,
  ): Promise<any> {
    const result = await fetch(
      this.env.activeCluster.getImportEndpoint(serviceName, stage),
      {
        method: 'post',
        headers: {
          Authorization: `Bearer ${token}`,
          'Content-Type': 'application/json',
        },
        body: exportData,
      },
    )

    const text = await result.text()
    try {
      return JSON.parse(text)
    } catch (e) {
      throw new Error(text)
    }
  }

  async reset(
    serviceName: string,
    stage: string,
    token?: string,
  ): Promise<void> {
    const result = await fetch(
      this.env.activeCluster.getApiEndpoint(serviceName, stage),
      {
        method: 'post',
        headers: {
          Authorization: `Bearer ${token}`,
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          query: `mutation {
          resetData
        }`,
        }),
      },
    )

    const text = await result.text()
    try {
      return JSON.parse(text)
    } catch (e) {
      throw new Error(text)
    }
  }

  async addProject(
    name: string,
    stage: string,
    secrets: string[] | null,
  ): Promise<SimpleProjectInfo> {
    const mutation = `\
      mutation addProject($name: String! $stage: String! $secrets: [String!]) {
        addProject(input: {
          name: $name,
          stage: $stage
          secrets: $secrets
        }) {
          project {
            name
          }
        }
      }
      `

    const { addProject: { project } } = await this.client.request<{
      addProject: { project: SimpleProjectInfo }
    }>(mutation, {
      name,
      stage,
      secrets,
    })

    // TODO set project definition, should be possibility in the addProject mutation

    return project
  }

  async deploy(
    name: string,
    stage: string,
    types: string,
    dryRun: boolean,
    secrets: string[] | null,
  ): Promise<any> {
    // TODO add dryRun to query as soon as the backend is ready
    const mutation = `\
      mutation($name: String!, $stage: String! $types: String! $dryRun: Boolean $secrets: [String!]) {
        deploy(input: {
          name: $name
          stage: $stage
          types: $types
          dryRun: $dryRun
          secrets: $secrets
        }) {
          errors {
            type
            field
            description
          }
          migration {
            ...MigrationFragment
          }
        }
      }
      ${MIGRATION_FRAGMENT}
    `

    const { deploy } = await this.client.request<{
      deploy: DeployPayload
    }>(mutation, {
      name,
      stage,
      types,
      dryRun,
      secrets,
    })

    return deploy
  }

  async listProjects(): Promise<Project[]> {
    const { listProjects } = await this.client.request<{
      listProjects: Project[]
    }>(`
      {
        listProjects {
          name
          stage
        }
      }
    `)

    return listProjects
  }

  async getProject(name: string, stage: string): Promise<Project | null> {
    const { project } = await this.client.request<{
      project: Project | null
    }>(
      `
      query($name: String! $stage: String!) {
        project(name: $name stage: $stage) {
          name
          stage
        }
      }
    `,
      { name, stage },
    )

    return project
  }

  async getCluster(name: string, stage: string): Promise<Cluster | null> {
    const foundClusters: Cluster[] = []
    for (const cluster of this.env.clusters) {
      const client = this.getClient(cluster)
      try {
        const { project } = await client.request<{
          project: Project
        }>(
          `
            query ($name: String! $stage: String!) {
              project (name: $name stage: $stage) {
                name
                stage
              }
            }
          `,
          {
            name,
            stage,
          },
        )

        if (project) {
          foundClusters.push(cluster)
        }
      } catch (e) {
        //
      }
    }

    if (foundClusters.length === 1) {
      return foundClusters[0]
    }

    if (foundClusters.length > 1) {
      const clusterNames = foundClusters.map(c => c.name).join(', ')
      throw new Error(
        `The service name / stage combination "${name}@${stage}" is ambigious. It exists in clusters ${clusterNames}`,
      )
    }

    return null
  }

  async getClusterSafe(name: string, stage: string): Promise<Cluster> {
    const cluster = await this.getCluster(name, stage)
    if (!cluster) {
      throw new Error(
        `No cluster for "${name}@${stage}" found. Please make sure to deploy the stage ${stage}`,
      )
    }

    return cluster
  }

  async waitForLocalDocker(endpoint: string): Promise<void> {
    // dont send any auth information when running the authenticateCustomer mutation
    let valid = false
    while (!valid) {
      try {
        debug('requesting', endpoint)
        await request(
          endpoint,
          `
            {
              __schema {
                directives {
                  description
                }
              }
            }
            `,
        )
        valid = true
      } catch (e) {
        debug(e)
        valid = false
      }

      await new Promise(r => setTimeout(r, 500))
    }
  }

  async waitForMigration(
    name: string,
    stage: string,
    revision: number,
  ): Promise<void> {
    while (true) {
      const { migrationStatus } = await this.client.request<any>(
        `query ($name: String! $stage: String!) {
              migrationStatus(name: $name stage: $stage) {
                revision
                hasBeenApplied
              }
            }
        `,

        {
          stage,
          name,
        },
      )

      if (
        migrationStatus.revision >= revision &&
        migrationStatus.hasBeenApplied
      ) {
        return
      }

      await new Promise(r => setTimeout(r, 500))
    }
  }

  async authenticateCustomer(
    endpoint: string,
    token: string,
  ): Promise<AuthenticateCustomerPayload> {
    // dont send any auth information when running the authenticateCustomer mutation
    const result = await request<{
      authenticateCustomer: AuthenticateCustomerPayload
    }>(
      endpoint,
      `
      mutation ($token: String!) {
        authenticateCustomer(input: {
          auth0IdToken: $token
        }) {
          token
          user {
            id
          }
        }
      }
      `,
      { token },
    )

    return result.authenticateCustomer
  }
}

// only make this available in test mode
// if (process.env.NODE_ENV === 'test') {
//   Client.prototype.mock = function({ request, response }) {
//     if (!this.mocks) {
//       this.mocks = {}
//     }
//     this.mocks[JSON.stringify(request, null, 2)] = response
//   }
// }

function normalizeName(name: string) {
  return name.toLowerCase().trim()
}
