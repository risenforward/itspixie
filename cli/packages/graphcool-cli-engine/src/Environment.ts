import * as yaml from 'js-yaml'
import * as path from 'path'
import fs from './fs'
import { Output } from './Output/index'
import { Client } from './Client/Client'
import { Config } from './Config'
import { EnvDoesntExistError } from './errors/EnvDoesntExistError'
import { RC, Target, Targets } from './types/rc'
import { mapValues, merge } from 'lodash'
import { Args } from './types/common'
import Variables from './ProjectDefinition/Variables'
const debug = require('debug')('environment')

export class Environment {
  localRC: RC = {}
  globalRC: RC = {}
  out: Output
  client: Client
  config: Config
  args: Args

  constructor(out: Output, config: Config, client: Client) {
    this.out = out
    this.config = config
    this.client = client
    this.migrateOldFormat()
  }
  private setTestToken() {
    debug('taking graphcool test token')
    this.globalRC.platformToken = process.env.GRAPHCOOL_TEST_TOKEN!
  }

  get rc(): RC {
    return merge({}, this.globalRC, this.localRC)
  }

  get default(): Target | null {
    if (this.rc.targets && this.rc.targets.default) {
      return this.rc.targets.default
    }

    return null
  }

  /**
   * This is used to migrate the old .graphcool and .graphcoolrc to the new format
   */
  migrateOldFormat() {}
  // public loadDotGraphcool() {
  //   if (this.dotGraphcoolFilePath && fs.existsSync(this.dotGraphcoolFilePath)) {
  //     const configContent = fs.readFileSync(
  //       this.dotGraphcoolFilePath,
  //       'utf-8',
  //     )
  //     let file: any = null
  //     try {
  //       file = JSON.parse(configContent)
  //     } catch (e) {
  //       try {
  //         file = yaml.safeLoad(configContent)
  //       } catch (e) {
  //         this.out.error(`Could not load ${this.dotGraphcoolFilePath}. It's neither valid json nor valid yaml.`)
  //       }
  //     }
  //
  //     if (file.token) {
  //       debug(`loading .graphcool file: no token existing`)
  //       this.token = file.token
  //     }
  //
  //     if (file.targets) {
  //       debug(`loading .graphcool file: no targets existing`)
  //       this.targets = file.targets
  //     }
  //   }
  // }

  async loadYaml(file: string | null, filePath: string | null = null): any {
    if (file) {
      let content
      try {
        content = yaml.safeLoad(file)
      } catch (e) {
        this.out.error(`Yaml parsing error in ${filePath}: ${e.message}`)
      }
      const variables = new Variables(this.out, filePath || 'no filepath provided', this.args)
      content = await variables.populateJson(content)

      return content
    } else {
      return {}
    }
  }

  async load(args: Args) {
    const localFile = this.config.localRCPath && fs.pathExistsSync(this.config.localRCPath) ? fs.readFileSync(this.config.localRCPath, 'utf-8') : null
    const globalFile = this.config.globalRCPath && fs.pathExistsSync(this.config.globalRCPath) ? fs.readFileSync(this.config.globalRCPath, 'utf-8') : null

    await this.loadRCs(localFile, globalFile, args)

    if (process.env.NODE_ENV === 'test') {
      this.setTestToken()
    }
  }

  async loadRCs(localFile: string | null, globalFile: string | null, args: Args = {}): Promise<void> {
    this.args = args

    const localJson = await this.loadYaml(localFile, this.config.localRCPath)
    const globalJson = await this.loadYaml(globalFile, this.config.globalRCPath)

    this.deserializeRCs(localJson, globalJson, this.config.localRCPath, this.config.globalRCPath)
  }

  deserializeRCs(localFile: any, globalFile: any, localFilePath: string | null, globalFilePath: string | null): void {
    let allTargets = {...localFile.targets, ...globalFile.targets}

    // 1. resolve aliases
    // global is not allowed to access local variables
    globalFile.targets = this.resolveTargetAliases(globalFile.targets, globalFile.targets)

    // repeat this 2 times as potentially there could be a deeper indirection
    for(let i = 0; i < 2; i++) {
      // first resolve all aliases
      localFile.targets = this.resolveTargetAliases(localFile.targets, allTargets)

      allTargets = {...localFile.targets, ...globalFile.targets}
    }

    // at this point there should only be targets in the form of shared-eu-west-1/cj862nxg0000um3t0z64ls08
    // 2. convert cluster/id to Target
    localFile.targets = this.deserializeTargets(localFile.targets, localFilePath)
    globalFile.targets = this.deserializeTargets(globalFile.targets, globalFilePath)
    // check if clusters exist
    const allClusters = [...this.config.sharedClusters, ...Object.keys(globalFile.clusters || {}), ...Object.keys(localFile.clusters || {})]
    this.checkClusters(localFile.targets, allClusters, localFilePath)
    this.checkClusters(globalFile.targets, allClusters, globalFilePath)
    this.localRC = localFile
    this.globalRC = globalFile
  }

  checkClusters(targets: Targets, clusters: string[], filePath: string | null) {
    Object.keys(targets).forEach(key => {
      const target = targets[key]
      if (!clusters.includes(target.cluster)) {
        this.out.error(`Could not find cluster ${target.cluster} defined for target ${key} in ${filePath}`)
      }
    })
  }

  deserializeTargets(targets: {[key: string]: string}, filePath: string | null): Targets {
    return mapValues<string, Target>(targets, target => this.deserializeTarget(target, filePath))
  }

  deserializeTarget(target: string, filePath: string | null): Target {
    const splittedTarget = target.split('/')
    if (splittedTarget.length === 1) {
      this.out.error(`Could not parse target ${target} in ${filePath}`)
    }
    return {
      cluster: splittedTarget[0],
      id: splittedTarget[1]
    }
  }

  resolveTargetAliases = (targets, allTargets) => mapValues(targets, target =>
    this.isTargetAlias(target) ? this.resolveTarget(target, allTargets) : target
  )
  isTargetAlias = (target: string): boolean => target.split('/').length === 1
  resolveTarget = (
    target: string,
    targets: { [key: string]: string },
  ): string =>
    targets[target] ? this.resolveTarget(targets[target], targets) : target

  serializeRC(rc: RC): string {
    const copy: any = {...rc}
    if (copy.targets) {
      copy.targets = this.serializeTargets(copy.targets)
    }
    return yaml.safeDump(copy)
  }

  serializeTargets(targets: Targets) {
    return mapValues<Target, string>(targets, t => `${t.cluster}/${t.id}`)
  }

  setToken(token: string | undefined) {
    this.globalRC.platformToken = token
  }

  saveLocalRC() {

  }

  saveGlobalRC() {
    const file = this.serializeRC(this.globalRC)
    fs.writeFileSync(this.config.globalRCPath, file)
  }
}
