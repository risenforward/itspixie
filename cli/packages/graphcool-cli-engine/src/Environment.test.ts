import {Environment} from './Environment'
import {Output} from './Output'
import {Config} from './Config'
import {Client} from './Client/Client'

// lets write some tests

function makeEnvironment() {
  const config = new Config()
  const out = new Output(config)
  const client = new Client(config)
  return new Environment(out, config, client)
}

describe('config', () => {
  test('should work without global RC', async () => {
    const env = makeEnvironment()
    const localFile = `
platformToken: 'secret-token'
targets:
  dev: local/asdasd123
clusters:
  local:
    host: http://localhost:60000
    token: asdf
`

    debugger
    env.loadRCs(localFile, null)
    expect(env.localRC).toMatchSnapshot()
  })
  // test('should access targets from the global RC', async () => {
  // })
  // test('should override targets from the global RC', async () => {
  // })
})
