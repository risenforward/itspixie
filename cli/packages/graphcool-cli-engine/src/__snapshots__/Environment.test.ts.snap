// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`config should access targets from the global RC 1`] = `
Object {
  "clusters": Object {
    "local": Object {
      "host": "http://localhost:60000",
      "token": "asdf",
    },
  },
  "platformToken": "secret-token",
  "targets": Object {
    "dev": Object {
      "cluster": "local",
      "id": "asdasd123",
    },
    "test-alias": Object {
      "cluster": "local",
      "id": "asdasd123",
    },
  },
}
`;

exports[`config should override global target 1`] = `
Object {
  "clusters": Object {
    "local": Object {
      "host": "http://localhost:60000",
      "token": "asdf",
    },
  },
  "platformToken": "secret-token",
  "targets": Object {
    "dev": Object {
      "cluster": "shared-eu-west-1",
      "id": "asdfasdf",
    },
  },
}
`;

exports[`config should resolve a target alias refering after 1`] = `
Object {
  "clusters": Object {
    "local": Object {
      "host": "http://localhost:60000",
      "token": "asdf",
    },
  },
  "platformToken": "secret-token",
  "targets": Object {
    "dev": Object {
      "cluster": "local",
      "id": "asdasd123",
    },
    "test-alias": Object {
      "cluster": "local",
      "id": "asdasd123",
    },
  },
}
`;

exports[`config should resolve a target alias refering before 1`] = `
Object {
  "clusters": Object {
    "local": Object {
      "host": "http://localhost:60000",
      "token": "asdf",
    },
  },
  "platformToken": "secret-token",
  "targets": Object {
    "dev": Object {
      "cluster": "local",
      "id": "asdasd123",
    },
    "test-alias": Object {
      "cluster": "local",
      "id": "asdasd123",
    },
  },
}
`;

exports[`config should resolve target from the global RC 1`] = `
Object {
  "clusters": Object {
    "local": Object {
      "host": "http://localhost:60000",
      "token": "asdf",
    },
  },
  "platformToken": "secret-token",
  "targets": Object {
    "dev": Object {
      "cluster": "local",
      "id": "asdasd123",
    },
    "test-alias": Object {
      "cluster": "local",
      "id": "asdasd123",
    },
  },
}
`;

exports[`config should work without global RC 1`] = `
Object {
  "clusters": Object {
    "local": Object {
      "host": "http://localhost:60000",
      "token": "asdf",
    },
  },
  "platformToken": "secret-token",
  "targets": Object {
    "dev": Object {
      "cluster": "local",
      "id": "asdasd123",
    },
  },
}
`;
